server/vehicle-persistence.lua

```diff
@@ -128,4 +139,31 @@ AddEventHandler('entityRemoved', function(entity)
         local passenger = passengers[i]
         SetPedIntoVehicle(passenger.ped, veh, passenger.seat)
     end
-end)
\ No newline at end of file
+end)
+
+if persistence == 1 then return end
+
+local spawned = false
+
+local function spawnVehicles()
+    spawned = true
+    local vehicles = exports.qbx_vehicles:GetPlayerVehicles({ states = 0 })
+    for _, vehicle in ipairs(vehicles) do
+        if not vehicle.coords then return end
+
+        local coords = vector3(vehicle.coords.x, vehicle.coords.y, vehicle.coords.z)
+        local playerVehicle = exports.qbx_vehicles:GetPlayerVehicle(vehicle.id)
+        if not playerVehicle then return end
+
+        local _, veh = qbx.spawnVehicle({ spawnSource = coords, model = playerVehicle.props.model, props = playerVehicle.props})
+        exports.qbx_core:EnablePersistence(veh)
+        Entity(veh).state:set('vehicleid', vehicle.id, false)
+        SetVehicleDoorsLocked(veh, 2)
+        SetEntityHeading(veh, vehicle.coords.w)
+    end
+end
+
+RegisterNetEvent('QBCore:Server:OnPlayerLoaded', function()
+    local players = exports.qbx_core:GetQBPlayers()
+    if not spawned and #players == 1 then spawnVehicles() end
+end)
```

Comment:
Would this work as an onResourceStart event? Why does there need to be at least 1 client in the server for this to work?

Reply:
My understanding is that CreateVehicleServerSetter blocks waiting for a client to be in one sync range, so I suspect it should work fine if there are no clients on the server; it should simply wait for a client to come in range.

Reply:
If we're going to precompute the vehicle type we should do so in shared/vehicles.lua

We'd just need to ensure that every car in the game is represented in that file, which would include any add-on vehicles that servers download.

What happens if you run the function with no players connected and then a player connects and enters one sync range of the vehicle spawn?

Reply:
@solareon @mafewtm @D4isDAVID Any thoughts on this?

Reply:
You're gonna have to explain what you want thoughts on

Reply:
If we should add types to shared/vehicles

Reply:
Why do we need types in the vehicles.lua? What purpose does it server that category, hash, or model does not or could not?

Reply:
CreateVehicleServerSetter needs the type, so it requires us to have a client and spawn a vehicle to find the type. Category wouldn't work because servers use that field to categorize vehicles into sports cars etc

Reply:
I mean if it *has* to be added then I guess. If that's what we wanna do then I'll do a pr. Just lmk.

Reply:
No I got it. I'm doing a couple more things to it anyway

Reply:
Any chance you could write a script for it so that it's easy for me to update my config? lol

Reply:
There's actually a pretty good argument for creating a separate table to hold this data so that people can easily upgrade. Otherwise merging a new field in with existing data is a big pain

Reply:
Our current state is we have config data mixed with non config data. So it makes sense to split out fields of shared/vehicles into data/vehicles.lua for fields that aren't meant to be opinionated

Reply:
I'm not a particular fan of having two mile long data tables for the same thing either